File: scripts/tts.py
# -*- coding: utf-8 -*-
"""Speak Bardejov dusk-walk copy.

Default engine is macOS `say` (Laura / Zuzana / Daniel / Zosia / …).
Optional neural voices:  pip install edge-tts

  python3 scripts/tts.py
  python3 scripts/tts.py --lang sk --what tour
  python3 scripts/tts.py --section gate
  python3 scripts/tts.py --text "Vstúpite do Bardejova."
  python3 scripts/tts.py --list
  python3 scripts/tts.py --engine edge --out public/audio/sk-top.mp3 --section top
"""
from __future__ import annotations

import argparse
import asyncio
import json
import re
import shutil
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
LANG_JS = ROOT / "src" / "lang.js"

LANGS = ("sk", "cs", "en", "pl", "hu", "uk")
VOICE_KEYS = ("top", "gate", "pathways", "lessons", "suburb", "walk", "eternity")

SAY_PREF = {
    "sk": ["laura", "zuzana"],
    "cs": ["iveta", "zuzana"],
    "en": ["daniel", "serena", "martha", "samantha"],
    "pl": ["zosia"],
    "hu": ["tünde", "tunde", "mariska"],
    "uk": ["lesya", "lesja"],
}
SAY_LOCALE = {
    "sk": "sk_",
    "cs": "cs_",
    "en": "en_gb",
    "pl": "pl_",
    "hu": "hu_",
    "uk": "uk_",
}
EDGE_VOICE = {
    "sk": "sk-SK-ViktoriaNeural",
    "cs": "cs-CZ-VlastaNeural",
    "en": "en-GB-SoniaNeural",
    "pl": "pl-PL-ZofiaNeural",
    "hu": "hu-HU-NoemiNeural",
    "uk": "uk-UA-PolinaNeural",
}
SKIP_SAY = re.compile(
    r"zarvox|whisper|bells|cellos|boing|bad news|good news|deranged|"
    r"trinoids|organ|wobble|jester|princess|junior|albert|bahh|bubbles|"
    r"superstar|kathy|fred",
    re.I,
)


def prepare_speech(text: str) -> str:
    line = str(text or "")
    line = re.sub(r"UNESCO|ЮНЕСКО", "Unesco", line)
    line = re.sub(r"[·•]", ", ", line)
    line = re.sub(r"[—–]", ", ", line)
    return re.sub(r"\s+", " ", line).strip()


def load_i18n() -> dict:
    if not LANG_JS.exists():
        sys.exit("missing src/lang.js — run: npm run i18n")
    raw = LANG_JS.read_text(encoding="utf-8")
    marker = "export const I18N = "
    start = raw.find(marker)
    if start < 0:
        sys.exit("could not parse src/lang.js")
    payload = raw[start + len(marker) :].strip()
    if payload.endswith(";"):
        payload = payload[:-1]
    return json.loads(payload)


def mac_voices() -> list[tuple[str, str]]:
    if not shutil.which("say"):
        return []
    out = subprocess.check_output(["say", "-v", "?"], text=True, errors="replace")
    rows = []
    for line in out.splitlines():
        m = re.match(r"^(\S+(?:\s+\S+)*?)\s+([a-z]{2}[_-][A-Za-z0-9]+)\s+#", line)
        if not m:
            continue
        name, loc = m.group(1).strip(), m.group(2).replace("-", "_").lower()
        if SKIP_SAY.search(name):
            continue
        rows.append((name, loc))
    return rows


def pick_say_voice(lang: str, override: str | None) -> str | None:
    if override:
        return override
    pref = SAY_PREF.get(lang, [])
    loc = SAY_LOCALE.get(lang, f"{lang}_")
    voices = mac_voices()
    scored: list[tuple[int, str]] = []
    for name, code in voices:
        s = -1
        if code.startswith(loc) or code == loc.rstrip("_"):
            s = 90
        elif lang == "en" and code.startswith("en_"):
            s = 40
        else:
            continue
        low = name.lower()
        if any(p in low for p in pref):
            s += 36
        scored.append((s, name))
    if not scored:
        return None
    scored.sort(reverse=True)
    return scored[0][1]


def speak_say(text: str, voice: str | None, rate: int, out: Path | None) -> None:
    cmd = ["say", "-r", str(rate)]
    if voice:
        cmd += ["-v", voice]
    if out:
        out.parent.mkdir(parents=True, exist_ok=True)
        cmd += ["-o", str(out)]
        if out.suffix.lower() in {".wav", ".wave"}:
            cmd += ["--data-format=LEF32@22050"]
    cmd.append(text)
    subprocess.run(cmd, check=True)


async def speak_edge(text: str, voice: str, rate: str, out: Path | None) -> None:
    try:
        import edge_tts
    except ImportError:
        sys.exit("edge-tts is not installed. Run:  pip3 install edge-tts")
    communicate = edge_tts.Communicate(text, voice, rate=rate)
    if out:
        out.parent.mkdir(parents=True, exist_ok=True)
        await communicate.save(str(out))
        return
    player = shutil.which("afplay") or shutil.which("ffplay")
    if not player:
        sys.exit("edge-tts needs --out FILE, or afplay/ffplay on PATH to play")
    tmp = Path("/tmp/bardejov-tts.mp3")
    await communicate.save(str(tmp))
    subprocess.run([player, str(tmp)], check=True)


def lines_for(pack: dict, what: str, section: str | None) -> list[str]:
    rows: list[str] = []
    if what in {"voice", "all"}:
        voice = pack.get("voice") or {}
        keys = [section] if section else list(VOICE_KEYS)
        for key in keys:
            if key not in voice:
                sys.exit(f"unknown section {key!r}. try: {', '.join(VOICE_KEYS)}")
            rows.append(voice[key])
    if what in {"tour", "all"}:
        if section and what == "tour":
            sys.exit("--section is for voice chapters, not tour")
        for step in pack.get("tour") or []:
            title = (step.get("title") or "").strip()
            k = (step.get("k") or "").strip()
            rows.append(f"{title}. {k}".strip() if k else title)
    return [prepare_speech(x) for x in rows if prepare_speech(x)]


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Text-to-speech for Bardejov dusk-walk copy.")
    p.add_argument("--lang", choices=LANGS, default="sk")
    p.add_argument("--what", choices=("voice", "tour", "all"), default="voice")
    p.add_argument("--section", choices=VOICE_KEYS, help="one voice chapter (default: all chapters)")
    p.add_argument("--text", help="speak this string instead of i18n copy")
    p.add_argument("--engine", choices=("say", "edge", "auto"), default="auto")
    p.add_argument("--voice", help="override voice name (say) or Edge short name")
    p.add_argument("--rate", type=int, default=165, help="say words-per-minute (default 165)")
    p.add_argument("--out", type=Path, help="write audio instead of playing")
    p.add_argument("--list", action="store_true", help="list macOS voices and exit")
    return p.parse_args()


def resolve_engine(name: str) -> str:
    if name != "auto":
        return name
    if shutil.which("say"):
        return "say"
    return "edge"


def main() -> None:
    args = parse_args()
    if args.list:
        voices = mac_voices()
        if not voices:
            sys.exit("macOS say is not available")
        for name, loc in voices:
            mark = ""
            for lid, prefix in SAY_LOCALE.items():
                if loc.startswith(prefix) or (lid == "en" and loc.startswith("en_")):
                    mark = f"  [{lid}]"
                    break
            print(f"{name:20} {loc}{mark}")
        return

    engine = resolve_engine(args.engine)
    if args.text:
        chunks = [prepare_speech(args.text)]
    else:
        packs = load_i18n()
        pack = packs.get(args.lang)
        if not pack:
            sys.exit(f"no i18n pack for {args.lang}")
        chunks = lines_for(pack, args.what, args.section)
    if not chunks:
        sys.exit("nothing to speak")

    if engine == "say":
        voice = pick_say_voice(args.lang, args.voice)
        if not voice:
            sys.exit(f"no macOS voice for {args.lang}. try --engine edge or --list")
        print(f"say · {args.lang} · {voice}", file=sys.stderr)
        if args.out and len(chunks) > 1:
            sys.exit("--out with say can write one file; pass --section or --text")
        for i, line in enumerate(chunks):
            print(line)
            speak_say(line, voice, args.rate, args.out if i == 0 else None)
        return

    voice = args.voice or EDGE_VOICE[args.lang]
    rate = f"{max(-50, min(50, args.rate - 175))}%"
    print(f"edge · {args.lang} · {voice}", file=sys.stderr)
    if args.out and len(chunks) > 1:
        sys.exit("--out with edge can write one file; pass --section or --text")
    joined = " ".join(chunks) if args.out else None
    if joined is not None:
        print(joined)
        asyncio.run(speak_edge(joined, voice, rate, args.out))
        return
    for line in chunks:
        print(line)
        asyncio.run(speak_edge(line, voice, rate, None))


if __name__ == "__main__":
    main()

File: scripts/tts.mjs
#!/usr/bin/env node
/* Speak Bardejov dusk-walk copy.

  npm run tts
  npm run tts -- --lang sk --what tour
  npm run tts -- --section gate
  npm run tts -- --text "Vstúpite do Bardejova."
  npm run tts -- --list
  npm run tts -- --engine edge --out public/audio/sk-top.mp3 --section top
*/
import { spawnSync } from 'node:child_process';
import { mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { dirname, extname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const root = join(dirname(fileURLToPath(import.meta.url)), '..');
const localesDir = join(root, 'locales');
const langs = JSON.parse(readFileSync(join(localesDir, 'langs.json'), 'utf8'));
const LANGS = langs.map((l) => l.id);
const VOICE_KEYS = ['top', 'gate', 'pathways', 'lessons', 'suburb', 'walk', 'eternity'];
const SAY_PREF = {
  sk: ['laura', 'zuzana'],
  cs: ['iveta', 'zuzana'],
  en: ['daniel', 'serena', 'martha', 'samantha'],
  pl: ['zosia'],
  hu: ['tünde', 'tunde', 'mariska'],
  uk: ['lesya', 'lesja'],
};
const SAY_LOCALE = {
  sk: 'sk_',
  cs: 'cs_',
  en: 'en_gb',
  pl: 'pl_',
  hu: 'hu_',
  uk: 'uk_',
};
const EDGE_VOICE = {
  sk: 'sk-SK-ViktoriaNeural',
  cs: 'cs-CZ-VlastaNeural',
  en: 'en-GB-SoniaNeural',
  pl: 'pl-PL-ZofiaNeural',
  hu: 'hu-HU-NoemiNeural',
  uk: 'uk-UA-PolinaNeural',
};
const SKIP_SAY =
  /zarvox|whisper|bells|cellos|boing|bad news|good news|deranged|trinoids|organ|wobble|jester|princess|junior|albert|bahh|bubbles|superstar|kathy|fred/i;

function fail(msg) {
  console.error(msg);
  process.exit(1);
}

function which(bin) {
  const finder = process.platform === 'win32' ? 'where' : 'which';
  const r = spawnSync(finder, [bin], { encoding: 'utf8' });
  if (r.status !== 0) return '';
  return (r.stdout || '').trim().split(/\r?\n/)[0] || '';
}

function prepareSpeech(text) {
  let line = String(text || '');
  line = line.replace(/UNESCO|ЮНЕСКО/g, 'Unesco');
  line = line.replace(/[·•]/g, ', ');
  line = line.replace(/[—–]/g, ', ');
  return line.replace(/\s+/g, ' ').trim();
}

function loadI18n() {
  const packs = {};
  for (const id of LANGS) {
    const path = join(localesDir, `${id}.json`);
    try {
      packs[id] = JSON.parse(readFileSync(path, 'utf8'));
    } catch {
      fail(`missing ${path}`);
    }
  }
  return packs;
}

function macVoices() {
  if (!which('say')) return [];
  const r = spawnSync('say', ['-v', '?'], { encoding: 'utf8' });
  if (r.status !== 0) return [];
  const rows = [];
  for (const line of (r.stdout || '').split(/\r?\n/)) {
    const m = line.match(/^(\S+(?:\s+\S+)*?)\s+([a-z]{2}[_-][A-Za-z0-9]+)\s+#/);
    if (!m) continue;
    const name = m[1].trim();
    const loc = m[2].replace(/-/g, '_').toLowerCase();
    if (SKIP_SAY.test(name)) continue;
    rows.push([name, loc]);
  }
  return rows;
}

function pickSayVoice(lang, override) {
  if (override) return override;
  const pref = SAY_PREF[lang] || [];
  const loc = SAY_LOCALE[lang] || `${lang}_`;
  const scored = [];
  for (const [name, code] of macVoices()) {
    let s = -1;
    if (code.startsWith(loc) || code === loc.replace(/_$/, '')) s = 90;
    else if (lang === 'en' && code.startsWith('en_')) s = 40;
    else continue;
    const low = name.toLowerCase();
    if (pref.some((p) => low.includes(p))) s += 36;
    scored.push([s, name]);
  }
  if (!scored.length) return null;
  scored.sort((a, b) => b[0] - a[0] || b[1].localeCompare(a[1]));
  return scored[0][1];
}

function speakSay(text, voice, rate, out) {
  const cmd = ['-r', String(rate)];
  if (voice) cmd.push('-v', voice);
  if (out) {
    mkdirSync(dirname(out), { recursive: true });
    cmd.push('-o', out);
    const ext = extname(out).toLowerCase();
    if (ext === '.wav' || ext === '.wave') cmd.push('--data-format=LEF32@22050');
  }
  cmd.push(text);
  const r = spawnSync('say', cmd, { stdio: 'inherit' });
  if (r.status !== 0) fail('say failed');
}

async function speakEdge(text, voice, rate, out) {
  let Communicate;
  try {
    ({ Communicate } = await import('@travisvn/edge-tts'));
  } catch {
    fail('edge TTS needs @travisvn/edge-tts. Run: npm i -D @travisvn/edge-tts');
  }
  const communicate = new Communicate(text, { voice, rate });
  const buffers = [];
  for await (const chunk of communicate.stream()) {
    if (chunk.type === 'audio' && chunk.data) buffers.push(chunk.data);
  }
  if (!buffers.length) fail('edge TTS returned no audio');
  const audio = Buffer.concat(buffers);
  if (out) {
    mkdirSync(dirname(out), { recursive: true });
    writeFileSync(out, audio);
    return;
  }
  const player = which('afplay') || which('ffplay');
  if (!player) fail('edge TTS needs --out FILE, or afplay/ffplay on PATH to play');
  const tmp = join(tmpdir(), 'bardejov-tts.mp3');
  writeFileSync(tmp, audio);
  const playArgs = player.endsWith('ffplay') ? ['-nodisp', '-autoexit', tmp] : [tmp];
  const r = spawnSync(player, playArgs, { stdio: 'inherit' });
  if (r.status !== 0) fail(`${player} failed`);
}

function linesFor(pack, what, section) {
  const rows = [];
  if (what === 'voice' || what === 'all') {
    const voice = pack.voice || {};
    const keys = section ? [section] : VOICE_KEYS;
    for (const key of keys) {
      if (!(key in voice)) fail(`unknown section '${key}'. try: ${VOICE_KEYS.join(', ')}`);
      rows.push(voice[key]);
    }
  }
  if (what === 'tour' || what === 'all') {
    if (section && what === 'tour') fail('--section is for voice chapters, not tour');
    for (const step of pack.tour || []) {
      const title = (step.title || '').trim();
      const k = (step.k || '').trim();
      rows.push(k ? `${title}. ${k}` : title);
    }
  }
  return rows.map(prepareSpeech).filter(Boolean);
}

function parseArgs(argv) {
  const args = {
    lang: 'sk',
    what: 'voice',
    section: null,
    text: null,
    engine: 'auto',
    voice: null,
    rate: 165,
    out: null,
    list: false,
    help: false,
  };
  const take = (flag, i) => {
    const v = argv[i + 1];
    if (v == null || v.startsWith('--')) fail(`missing value for ${flag}`);
    return v;
  };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--help' || a === '-h') args.help = true;
    else if (a === '--list') args.list = true;
    else if (a === '--lang') args.lang = take(a, i++);
    else if (a === '--what') args.what = take(a, i++);
    else if (a === '--section') args.section = take(a, i++);
    else if (a === '--text') args.text = take(a, i++);
    else if (a === '--engine') args.engine = take(a, i++);
    else if (a === '--voice') args.voice = take(a, i++);
    else if (a === '--rate') args.rate = Number(take(a, i++));
    else if (a === '--out') args.out = take(a, i++);
    else fail(`unknown flag ${a}`);
  }
  if (args.help) return args;
  if (!LANGS.includes(args.lang)) fail(`--lang must be ${LANGS.join(', ')}`);
  if (!['voice', 'tour', 'all'].includes(args.what)) fail('--what must be voice, tour or all');
  if (args.section && !VOICE_KEYS.includes(args.section)) {
    fail(`--section must be ${VOICE_KEYS.join(', ')}`);
  }
  if (!['say', 'edge', 'auto'].includes(args.engine)) fail('--engine must be say, edge or auto');
  if (!Number.isFinite(args.rate)) fail('--rate must be a number');
  return args;
}

function resolveEngine(name) {
  if (name !== 'auto') return name;
  return which('say') ? 'say' : 'edge';
}

function printHelp() {
  console.log(`Text-to-speech for Bardejov dusk-walk copy.

  npm run tts
  npm run tts -- --lang sk --what tour
  npm run tts -- --section gate
  npm run tts -- --text "Vstúpite do Bardejova."
  npm run tts -- --list
  npm run tts -- --engine edge --out public/audio/sk-top.mp3 --section top`);
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  if (args.help) {
    printHelp();
    return;
  }
  if (args.list) {
    const voices = macVoices();
    if (!voices.length) fail('macOS say is not available');
    for (const [name, loc] of voices) {
      let mark = '';
      for (const [lid, prefix] of Object.entries(SAY_LOCALE)) {
        if (loc.startsWith(prefix) || (lid === 'en' && loc.startsWith('en_'))) {
          mark = `  [${lid}]`;
          break;
        }
      }
      console.log(`${name.padEnd(20)} ${loc}${mark}`);
    }
    return;
  }

  const engine = resolveEngine(args.engine);
  const chunks = args.text
    ? [prepareSpeech(args.text)]
    : linesFor(
        loadI18n()[args.lang] || fail(`no i18n pack for ${args.lang}`),
        args.what,
        args.section
      );
  if (!chunks.length) fail('nothing to speak');

  if (engine === 'say') {
    const voice = pickSayVoice(args.lang, args.voice);
    if (!voice) fail(`no macOS voice for ${args.lang}. try --engine edge or --list`);
    console.error(`say · ${args.lang} · ${voice}`);
    if (args.out && chunks.length > 1)
      fail('--out with say can write one file; pass --section or --text');
    for (let i = 0; i < chunks.length; i++) {
      console.log(chunks[i]);
      speakSay(chunks[i], voice, args.rate, i === 0 ? args.out : null);
    }
    return;
  }

  const voice = args.voice || EDGE_VOICE[args.lang];
  const rate = `${Math.max(-50, Math.min(50, args.rate - 175))}%`;
  console.error(`edge · ${args.lang} · ${voice}`);
  if (args.out && chunks.length > 1)
    fail('--out with edge can write one file; pass --section or --text');
  if (args.out) {
    const joined = chunks.join(' ');
    console.log(joined);
    await speakEdge(joined, voice, rate, args.out);
    return;
  }
  for (const line of chunks) {
    console.log(line);
    await speakEdge(line, voice, rate, null);
  }
}

main().catch((err) => fail(err?.message || String(err)));
